fire/cli/ts/statistics.py

```diff
@@ -0,0 +1,337 @@
+        start = datetime.date(date.year, 1, 1).toordinal()
+        year_length = datetime.date(date.year + 1, 1, 1).toordinal() - start
+        dec_years.append(date.year + float(date.toordinal() - start) / year_length)
+    return np.array(dec_years).reshape(-1, 1)
+
+
+def reg_KI(
+    x_data: npt.NDArray[np.float64],
+    Y: npt.NDArray[np.float64],
+    y_data: npt.NDArray[np.float64],
+    statistic: float,
+    MS_res: List[float] = None,
+) -> tuple[List[List[float]], float]:
+    """
+    Produce a confidence interval for a linear fit
+
+    """
+    if not x_data.any():
+        raise ValueError("Datoer mangler")
+    if not Y.any():
+        raise ValueError("Fittede værdier mangler")
+    if not y_data.any():
+        raise ValueError("Højdemålinger mangler")
+
+    # setup
+    X = np.array([np.ones(len(x_data)), x_data.flatten()]).T  # kernel matrix
+    XTX_1 = np.linalg.inv(X.T @ X)  # (transpose(X)*X)^(-1)
+
+    if not MS_res:
+        # residual sum of squares
+        SS_res = sum(np.square(Y - y_data))
+        # MS_res (error variance)
+        MS_res = SS_res / (len(y_data) - 2)
+
+    # Method from:
+    # eq 10 p 134 of Statistical Models for Engineers (T. Tarpey)
+    warnings.filterwarnings("ignore")
+
+    # std. error
+    std_error = []
+    for x0i in x_data.flatten():  # for each date calculate std. error
+        x0 = np.array([1, x0i])  # date in vector form
+        std_error.append([np.sqrt(MS_res * x0 @ XTX_1 @ x0)])
+
+    # confidence intervals
+    std_error = np.array(std_error)
+    YL = Y - statistic * std_error
+    YU = Y + statistic * std_error
+
+    warnings.filterwarnings("default")
+
+    # std and conf int of slope
+    std_slope = np.sqrt((MS_res * XTX_1)[1, 1])
+    return [YL, YU], std_slope * 1e3
+
+
+def uplift_fit(
+    dates: npt.NDArray[np.float64], u: float, med: datetime.date, meh: float
+) -> npt.NDArray[np.float64]:
+    """
+    Fit the DTU uplift rate to the middle epoch height
+
+    """
+    if not dates.any():
+        raise ValueError("Indtast datoer for fit")
+    if not u:
+        raise ValueError("Indtast DTU uplift hældning")
+    if not med:
+        raise ValueError("Indtast middle epoch date")
+    if not meh:
+        raise ValueError("Indtast middle epoch height")
+
+    med = to_decimal_year([med]).flatten()[0]
+    # mm -> m, d -> y and compensate for MED
+    u = u / 1e3
+    epoch = [(date - med) for date in dates.flatten()]
+    # fit uplift to intercept the middle epoch height)
+    uplift_corrected = np.array([i * u + meh for i in epoch])
+    return uplift_corrected
+
+
+def convert_UTM2lonlat(
+    E: npt.NDArray[np.float64], N: npt.NDArray[np.float64]
+) -> tuple[npt.NDArray[np.float64], npt.NDArray[np.float64]]:
+    """
+    Convert UTM easting, norting to lon, lat
+
+    """
+    if not E.any() or not N.any():
+        raise ValueError("Indtast Easting og Northing koordinater")
+
+    myProj = Proj(proj="utm", zone=32, ellps="WGS84", preserve_units=True)
```

Comment:
myProj kan være en global konstant: Det er ikke nødvendigt at reinitialisere den for hvert kald. Løft den et par linjer op og sig:
```python
UTM32 = Proj(proj="utm", zone=32, ellps="WGS84", preserve_units=True)
...
return   UTM32(E, N, inverse=True)
```

Reply:
God idé! Jeg har rettet min kode i gnss.py efter dette foreslag.